source/language/openpulse.rst

```diff
@@ -0,0 +1,317 @@
+
+   // Define a constant pulse of length l
+   constant(l:length)
+
+Play instruction
+----------------
+
+Pulses are scheduled using the ``play`` instruction. These instructions may
+only appear inside a ``defcal`` block!
+
+Play instructions have two required parameters:
+
+- a value of type ``pulse`` representing the pulse envelope
+- the frame to use for the pulse
+
+.. code-block:: none
+
+   // Play a 3 sample pulse on qubit 0's "drive" frame
+   play([1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j]) frameof("drive", %0);
+
+   // Play a gaussian on qubit 1's "drive" frame
+   frame f1 = frameof("drive", %1);
+   play(gaussian(...)) f1;
+
+Capture Instruction
+-------------------
+
+Acquisition is scheduled by a ``capture`` instruction. This is a special
+``kernel`` function which is specified by a hardware vendor. The measurement
+process is difficult to describe generically due to the wide variety of
+hardware and measurement methods. Like the play instruction, these instructions
+may only appear inside a ``defcal`` block!
+
+The only required parameter is a ``frame``.
+
+The following are possible parameters that might be included:
+
+- A "duration" of type ``length``, if it cannot be inferred from other parameters
+- A "filter", which is dot product-ed with the measured IQ the distill the
+  result into a single IQ value
+- A "tag", which could be used to identify which branch of an if statement was
+  traversed
+
+Again it is up to the hardware vendor to determine the parameters and write a
+kernel definition at the top-level, such as:
+
+.. code-block:: none
+
+   // Minimum requirement
+   kernel capture(frame output) -> complex[32];
+
+   // A capture command with more features
+   kernel capture(frame output, pulse filter) -> complex[32];
+
+The return type of a ``capture`` command varies. It could be a raw trace, ie. a
+list of samples taken over a short period of time. It could be some averaged IQ
+value. It could be a classified bit. Or it could even have no return value,
+pushing the results into some buffer which is then accessed outside the program.
+
+Timing
+------
+
+Each frame maintains its own "clock". When a pulse is played the clock for 
+that frame advances by the length of the pulse. 
+
+For frames, everything behaves analogous to qubits in the 
+`Delays <delays.html>`_ section of this specification. There are however some
+small differences.
+
+The ``delay`` instruction may take a frame instead of a qubit. The ``barrier`` 
+instruction may also take a list of frames intead of a list of qubits.
+
+``defcal`` blocks have an implicit barrier on every frame used within the block,
+meaning that clocks are guaranteed to be aligned at the start of the block.
+These blocks also need to have a well-defined length, similar to the ``boxas``
+block.
+
+.. code-block:: none
+
+   pulse p = ...; // some 100dt pulse
+
+   defcal simultaneous_pulsed_gate %0 {
+     play(p) frameof("drive0", %0);
+     delay[20dt] frameof("drive1", %0);
```

Comment:
I think this syntax matches the gate syntax, right? We don't allow other QASM gates within `defcal`s, do we? Perhaps we could keep the syntax for `delay` the same as the syntax for `play`? `play(arg) frame;` and `delay(arg) frame;`?

Reply:
Relates to this issue: https://github.com/Qiskit/openqasm/issues/121